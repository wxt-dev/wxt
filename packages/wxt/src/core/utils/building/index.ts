export * from './build-entrypoints';
export * from './detect-dev-changes';
export * from './find-entrypoints';
export * from './group-entrypoints';
export * from './internal-build';
export * from './rebuild';
