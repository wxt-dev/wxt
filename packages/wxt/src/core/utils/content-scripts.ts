import type { Browser } from '@wxt-dev/browser';
import { ContentScriptEntrypoint, ResolvedConfig } from '../../types';
import { getEntrypointBundlePath } from './entrypoints';
import { ManifestContentScript } from './types';

/**
 * Returns a unique and consistent string hash based on a content scripts options.
 *
 * It is able to recognize default values,
 */
export function hashContentScriptOptions(
  options: ContentScriptEntrypoint['options'],
): string {
  const simplifiedOptions = mapWxtOptionsToContentScript(
    options,
    undefined,
    undefined,
  );

  // Remove undefined fields and use defaults to generate hash
  Object.keys(simplifiedOptions).forEach((key) => {
    // @ts-expect-error: key not typed as keyof ...
    if (simplifiedOptions[key] == null) delete simplifiedOptions[key];
  });

  const withDefaults: ManifestContentScript = {
    exclude_globs: [],
    exclude_matches: [],
    include_globs: [],
    match_about_blank: false,
    run_at: 'document_idle',
    all_frames: false,
    // @ts-expect-error: Untyped
    match_origin_as_fallback: false,
    world: 'ISOLATED',
    ...simplifiedOptions,
  };
  return JSON.stringify(
    Object.entries(withDefaults)
      // Sort any arrays so their values are consistent
      .map<[string, unknown]>(([key, value]) => {
        if (Array.isArray(value)) return [key, value.sort()];
        else return [key, value];
      })
      // Sort all the fields alphabetically
      .sort((l, r) => l[0].localeCompare(r[0])),
  );
}

export function mapWxtOptionsToContentScript(
  options: ContentScriptEntrypoint['options'],
  js: string[] | undefined,
  css: string[] | undefined,
): ManifestContentScript {
  return {
    matches: options.matches ?? [],
    all_frames: options.allFrames,
    match_about_blank: options.matchAboutBlank,
    exclude_globs: options.excludeGlobs,
    exclude_matches: options.excludeMatches,
    include_globs: options.includeGlobs,
    run_at: options.runAt,
    css,
    js,

    // @ts-expect-error: Untyped
    match_origin_as_fallback: options.matchOriginAsFallback,
    world: options.world,
  };
}

export function mapWxtOptionsToRegisteredContentScript(
  options: ContentScriptEntrypoint['options'],
  js: string[] | undefined,
  css: string[] | undefined,
): Omit<Browser.scripting.RegisteredContentScript, 'id'> {
  return {
    allFrames: options.allFrames,
    excludeMatches: options.excludeMatches,
    matches: options.matches,
    runAt: options.runAt,
    js,
    css,
    world: options.world,
  };
}

export function getContentScriptJs(
  config: ResolvedConfig,
  entrypoint: ContentScriptEntrypoint,
): string[] {
  return [getEntrypointBundlePath(entrypoint, config.outDir, '.js')];
}
