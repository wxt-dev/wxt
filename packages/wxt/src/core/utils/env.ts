import { config } from 'dotenv';
import { expand } from 'dotenv-expand';
import type { TargetBrowser } from '../../types';

/**
 * Load environment files based on the current mode and browser.
 */
export function loadEnv(mode: string, browser: TargetBrowser) {
  return expand(
    config({
      // Files on top override files below
      path: [
        `.env.${mode}.${browser}.local`,
        `.env.${mode}.${browser}`,
        `.env.${browser}.local`,
        `.env.${browser}`,
        `.env.${mode}.local`,
        `.env.${mode}`,
        `.env.local`,
        `.env`,
      ],
    }),
  );
}
