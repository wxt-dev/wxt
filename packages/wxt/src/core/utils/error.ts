import { isAbsolute, relative, resolve } from 'node:path';
import pc from 'picocolors';
import { wxt } from '../wxt';

export interface BabelSyntaxError extends SyntaxError {
  code: 'BABEL_PARSER_SYNTAX_ERROR';
  frame?: string;
  id: string;
  loc: { line: number; column: number };
}

export function isBabelSyntaxError(
  error: SyntaxError & { code?: string },
): error is BabelSyntaxError {
  return error.code === 'BABEL_PARSER_SYNTAX_ERROR';
}

export function logBabelSyntaxError(error: BabelSyntaxError) {
  let filename = relative(wxt.config.root, error.id);
  if (filename.startsWith('..')) {
    filename = error.id;
  }
  let message = error.message.replace(
    /\(\d+:\d+\)$/,
    `(${filename}:${error.loc.line}:${error.loc.column + 1})`,
  );
  if (error.frame) {
    message += '\n\n' + pc.red(error.frame);
  }
  wxt.logger.error(message);
}

export interface ModuleNotFoundError extends Error {
  code: 'ERR_MODULE_NOT_FOUND';
}

export function isModuleNotFoundError(
  error: Error & { code?: string },
): error is ModuleNotFoundError {
  return error.code === 'ERR_MODULE_NOT_FOUND';
}

export interface ModuleNotFoundErrorDetails {
  id: string;
  importer: string;
}

export function parseModuleNotFoundError(
  error: ModuleNotFoundError,
): ModuleNotFoundErrorDetails | null {
  // Cannot find module 'xyz' imported from '/@fs//path/to/abc.ts'
  const match = error.message.match(
    /^Cannot find module '(.*)' imported from '(.*)'/,
  );
  return (
    match && {
      id: match[1],
      importer: match[2],
    }
  );
}

export interface VitePluginError extends Error {
  code: 'PLUGIN_ERROR';
  pluginCode: string;
  plugin: string;
  hook: string;
}

export function isVitePluginError(
  error: Error & { code?: string },
): error is VitePluginError {
  return error.code === 'PLUGIN_ERROR';
}

export interface ViteLoadFallbackError extends VitePluginError {
  plugin: 'vite:load-fallback';
  hook: 'load';
  path: string;
}

export function isViteLoadFallbackError(
  error: VitePluginError,
): error is ViteLoadFallbackError {
  return error.plugin === 'vite:load-fallback' && error.hook === 'load';
}

export function resolveErrorFiles(err: Error): string[] {
  const errorFiles: string[] = [];

  if ('cause' in err && err.cause instanceof Error) {
    err = err.cause;
  }

  // TODO: support syntax errors from ESBuild and others.
  if (err instanceof SyntaxError) {
    if (isBabelSyntaxError(err)) {
      logBabelSyntaxError(err);
      errorFiles.push(err.id);

      wxt.logger.info('Waiting for syntax error to be fixed...');
    }
  } else if (isModuleNotFoundError(err)) {
    const details = parseModuleNotFoundError(err);
    if (details && details.importer.startsWith('/@fs/')) {
      errorFiles.push(details.importer.slice(5));

      wxt.logger.error(err.message);
      wxt.logger.info('Waiting for import specifier to be fixed...');
    }
  } else if (
    isVitePluginError(err) &&
    isViteLoadFallbackError(err) &&
    err.pluginCode === 'ENOENT'
  ) {
    // Extract the importer from the error message, which is produced by Rollup:
    // https://github.com/rollup/rollup/blob/384d5333fbc3d8918b41856822376da2a65ccaa3/src/ModuleLoader.ts#L288-L289
    const importerMatch = err.message.match(
      // "Could not load /path/to/xyz.ts (imported by src/module.ts)"
      /\(imported by (.*)\)/,
    );
    if (importerMatch) {
      errorFiles.push(resolve(importerMatch[1]));
    }
    // Ignore non-absolute paths, which are likely virtual modules.
    if (isAbsolute(err.path)) {
      errorFiles.push(err.path);
    }
    if (errorFiles.length) {
      // The error was already logged by Vite, so just log this.
      wxt.logger.info('Waiting for missing file to be added...');
    }
  } else if (
    // Vite throws an error when Rollup cannot resolve an import at build time:
    // https://github.com/vitejs/vite/blob/998303b438734e8219715fe6883b97fb10404c16/packages/vite/src/node/build.ts#L1027-L1032
    err.message.includes('Rollup failed to resolve import')
  ) {
    const importerMatch = err.message.match(/ from "(.*)"/);
    if (importerMatch) {
      errorFiles.push(importerMatch[1]);

      // The error was already logged by Vite, so just log this.
      wxt.logger.info('Waiting for import specifier to be fixed...');
    }
  }

  return errorFiles;
}
