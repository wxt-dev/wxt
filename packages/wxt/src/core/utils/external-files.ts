import { BuildOutput } from '../../types';
import path from 'node:path';
import fs from 'fs-extra';
import { wxt } from '../wxt';

/**
 * Analyzes the build output to find all external files (files outside the project directory)
 * that are imported by the extension and should be included in the sources zip.
 */
export async function gatherExternalFiles(
  output: BuildOutput,
): Promise<string[]> {
  const externalFiles = new Set<string>();
  const sourcesRoot = path.resolve(wxt.config.zip.sourcesRoot);

  // Iterate through all build steps and chunks to find external module dependencies
  for (const step of output.steps) {
    for (const chunk of step.chunks) {
      if (chunk.type === 'chunk') {
        // Check each module ID (dependency) in the chunk
        for (const moduleId of chunk.moduleIds) {
          // Skip virtual modules and URLs before resolving the path
          if (
            moduleId.startsWith('virtual:') ||
            moduleId.startsWith('http') ||
            moduleId.includes('node_modules') ||
            !path.isAbsolute(moduleId)
          ) {
            continue;
          }

          const normalizedModuleId = path.resolve(moduleId);

          // Only include files that are outside the sources root directory
          if (!normalizedModuleId.startsWith(sourcesRoot)) {
            try {
              await fs.access(normalizedModuleId);
              externalFiles.add(normalizedModuleId);
            } catch {
              wxt.logger.debug(
                `Skipping non-existent external file: ${normalizedModuleId}`,
              );
            }
          }
        }
      }
    }
  }

  const externalFilesArray = Array.from(externalFiles);

  if (externalFilesArray.length > 0) {
    wxt.logger.info(
      `Found ${externalFilesArray.length} external source files to include in zip`,
    );
    externalFilesArray.forEach((file) => {
      wxt.logger.debug(
        `  External file: ${path.relative(process.cwd(), file)}`,
      );
    });
  }

  return externalFilesArray;
}
