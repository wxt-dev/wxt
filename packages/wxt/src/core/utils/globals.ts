import { ResolvedConfig } from '../../types';

export function getGlobals(
  config: ResolvedConfig,
): Array<{ name: string; value: any; type: string }> {
  return [
    {
      name: 'MANIFEST_VERSION',
      value: config.manifestVersion,
      type: `2 | 3`,
    },
    {
      name: 'BROWSER',
      value: config.browser,
      type:
        config.targetBrowsers.length === 0
          ? 'string'
          : config.targetBrowsers.map((browser) => `"${browser}"`).join(' | '),
    },
    {
      name: 'CHROME',
      value: config.browser === 'chrome',
      type: `boolean`,
    },
    {
      name: 'FIREFOX',
      value: config.browser === 'firefox',
      type: `boolean`,
    },
    {
      name: 'SAFARI',
      value: config.browser === 'safari',
      type: `boolean`,
    },
    {
      name: 'EDGE',
      value: config.browser === 'edge',
      type: `boolean`,
    },
    {
      name: 'OPERA',
      value: config.browser === 'opera',
      type: `boolean`,
    },
    {
      name: 'COMMAND',
      value: config.command,
      type: `"build" | "serve"`,
    },
  ];
}

export function getEntrypointGlobals(entrypointName: string) {
  return [
    {
      name: 'ENTRYPOINT',
      value: entrypointName,
      type: `string`,
    },
  ];
}
