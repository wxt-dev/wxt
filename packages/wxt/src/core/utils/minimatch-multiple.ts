import { minimatch, MinimatchOptions } from 'minimatch';

/**
 * Run [`minimatch`](https://npmjs.com/package/minimatch) against multiple
 * patterns.
 *
 * Supports negated patterns, the order does not matter. If your `search` string
 * matches any of the negative patterns, it will return `false`.
 *
 * @example
 * ```ts
 * minimatchMultiple('a.json', ['*.json', '!b.json']); // => true
 * minimatchMultiple('b.json', ['*.json', '!b.json']); // => false
 * ```
 */
export function minimatchMultiple(
  search: string,
  patterns: string[] | undefined,
  options?: MinimatchOptions,
): boolean {
  if (patterns == null) return false;

  const negatePatterns: string[] = [];
  const positivePatterns: string[] = [];
  for (const pattern of patterns) {
    if (pattern[0] === '!') negatePatterns.push(pattern.slice(1));
    else positivePatterns.push(pattern);
  }

  if (
    negatePatterns.some((negatePattern) =>
      minimatch(search, negatePattern, options),
    )
  )
    return false;

  return positivePatterns.some((positivePattern) =>
    minimatch(search, positivePattern, options),
  );
}
