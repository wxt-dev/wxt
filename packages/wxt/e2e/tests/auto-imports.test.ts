import { describe, it, expect } from 'vitest';
import { TestProject } from '../utils';
import spawn from 'nano-spawn';

describe('Auto Imports', () => {
  describe('imports: { ... }', () => {
    it('should generate a declaration file, imports.d.ts, for auto-imports', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare();

      expect(await project.serializeFile('.wxt/types/imports.d.ts'))
        .toMatchInlineSnapshot(`
          ".wxt/types/imports.d.ts
          ----------------------------------------
          // Generated by wxt
          export {}
          declare global {
            const ContentScriptContext: typeof import('wxt/utils/content-script-context')['ContentScriptContext']
            const InvalidMatchPattern: typeof import('wxt/utils/match-patterns')['InvalidMatchPattern']
            const MatchPattern: typeof import('wxt/utils/match-patterns')['MatchPattern']
            const MigrationError: typeof import('wxt/storage')['MigrationError']
            const browser: typeof import('wxt/browser')['browser']
            const createIframeUi: typeof import('wxt/utils/content-script-ui/iframe')['createIframeUi']
            const createIntegratedUi: typeof import('wxt/utils/content-script-ui/integrated')['createIntegratedUi']
            const createShadowRootUi: typeof import('wxt/utils/content-script-ui/shadow-root')['createShadowRootUi']
            const defineAppConfig: typeof import('wxt/utils/define-app-config')['defineAppConfig']
            const defineBackground: typeof import('wxt/utils/define-background')['defineBackground']
            const defineContentScript: typeof import('wxt/utils/define-content-script')['defineContentScript']
            const defineUnlistedScript: typeof import('wxt/utils/define-unlisted-script')['defineUnlistedScript']
            const defineWxtPlugin: typeof import('wxt/utils/define-wxt-plugin')['defineWxtPlugin']
            const fakeBrowser: typeof import('wxt/testing')['fakeBrowser']
            const injectScript: typeof import('wxt/utils/inject-script')['injectScript']
            const storage: typeof import('wxt/storage')['storage']
            const useAppConfig: typeof import('wxt/utils/app-config')['useAppConfig']
          }
          "
        `);
    });

    it('should include auto-imports in the project', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare();

      expect(await project.serializeFile('.wxt/wxt.d.ts'))
        .toMatchInlineSnapshot(`
          ".wxt/wxt.d.ts
          ----------------------------------------
          // Generated by wxt
          /// <reference types="wxt/vite-builder-env" />
          /// <reference path="./types/paths.d.ts" />
          /// <reference path="./types/i18n.d.ts" />
          /// <reference path="./types/globals.d.ts" />
          /// <reference path="./types/imports-module.d.ts" />
          /// <reference path="./types/imports.d.ts" />
          "
        `);
    });

    it('should generate the #imports module', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);
      // Project auto-imports should also be present
      project.addFile(
        'utils/time.ts',
        `export function startOfDay(date: Date): Date {
          throw Error("TODO")
        }`,
      );

      await project.prepare();

      expect(await project.serializeFile('.wxt/types/imports-module.d.ts'))
        .toMatchInlineSnapshot(`
          ".wxt/types/imports-module.d.ts
          ----------------------------------------
          // Generated by wxt
          // Types for the #import virtual module
          declare module '#imports' {
            export { browser } from 'wxt/browser';
            export { MigrationError, storage } from 'wxt/storage';
            export { useAppConfig } from 'wxt/utils/app-config';
            export { ContentScriptContext } from 'wxt/utils/content-script-context';
            export { createIframeUi } from 'wxt/utils/content-script-ui/iframe';
            export { createIntegratedUi } from 'wxt/utils/content-script-ui/integrated';
            export { createShadowRootUi } from 'wxt/utils/content-script-ui/shadow-root';
            export { defineAppConfig } from 'wxt/utils/define-app-config';
            export { defineBackground } from 'wxt/utils/define-background';
            export { defineContentScript } from 'wxt/utils/define-content-script';
            export { defineUnlistedScript } from 'wxt/utils/define-unlisted-script';
            export { defineWxtPlugin } from 'wxt/utils/define-wxt-plugin';
            export { injectScript } from 'wxt/utils/inject-script';
            export { InvalidMatchPattern, MatchPattern } from 'wxt/utils/match-patterns';
            export { fakeBrowser } from 'wxt/testing';
            export { startOfDay } from '../utils/time';
          }
          "
        `);
    });
  });

  describe('imports: false', () => {
    it('should not generate a imports.d.ts file', async () => {
      const project = new TestProject();
      project.setConfigFileConfig({
        imports: false,
      });
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare();

      expect(await project.fileExists('.wxt/types/imports.d.ts')).toBe(false);
    });

    it('should only include imports-module.d.ts in the the project', async () => {
      const project = new TestProject();
      project.setConfigFileConfig({
        imports: false,
      });
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare();

      expect(
        await project.serializeFile('.wxt/wxt.d.ts'),
      ).toMatchInlineSnapshot(
        `
        ".wxt/wxt.d.ts
        ----------------------------------------
        // Generated by wxt
        /// <reference types="wxt/vite-builder-env" />
        /// <reference path="./types/paths.d.ts" />
        /// <reference path="./types/i18n.d.ts" />
        /// <reference path="./types/globals.d.ts" />
        /// <reference path="./types/imports-module.d.ts" />
        "
      `,
      );
    });

    it('should only generate the #imports module', async () => {
      const project = new TestProject();
      project.setConfigFileConfig({
        imports: false,
      });
      project.addFile('entrypoints/popup.html', `<html></html>`);
      // Project auto-imports should also be present
      project.addFile(
        'utils/time.ts',
        `export function startOfDay(date: Date): Date {
              throw Error("TODO")
            }`,
      );

      await project.prepare();

      expect(await project.serializeFile('.wxt/types/imports-module.d.ts'))
        .toMatchInlineSnapshot(`
          ".wxt/types/imports-module.d.ts
          ----------------------------------------
          // Generated by wxt
          // Types for the #import virtual module
          declare module '#imports' {
            export { browser } from 'wxt/browser';
            export { MigrationError, storage } from 'wxt/storage';
            export { useAppConfig } from 'wxt/utils/app-config';
            export { ContentScriptContext } from 'wxt/utils/content-script-context';
            export { createIframeUi } from 'wxt/utils/content-script-ui/iframe';
            export { createIntegratedUi } from 'wxt/utils/content-script-ui/integrated';
            export { createShadowRootUi } from 'wxt/utils/content-script-ui/shadow-root';
            export { defineAppConfig } from 'wxt/utils/define-app-config';
            export { defineBackground } from 'wxt/utils/define-background';
            export { defineContentScript } from 'wxt/utils/define-content-script';
            export { defineUnlistedScript } from 'wxt/utils/define-unlisted-script';
            export { defineWxtPlugin } from 'wxt/utils/define-wxt-plugin';
            export { injectScript } from 'wxt/utils/inject-script';
            export { InvalidMatchPattern, MatchPattern } from 'wxt/utils/match-patterns';
            export { fakeBrowser } from 'wxt/testing';
          }
          "
        `);
    });

    it('should still generate the #imports module', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);
      // Project auto-imports should also be exported
      project.addFile(
        'utils/time.ts',
        `export function startOfDay(date: Date): Date {
              throw Error("TODO")
            }`,
      );

      await project.prepare();

      expect(await project.serializeFile('.wxt/types/imports-module.d.ts'))
        .toMatchInlineSnapshot(`
          ".wxt/types/imports-module.d.ts
          ----------------------------------------
          // Generated by wxt
          // Types for the #import virtual module
          declare module '#imports' {
            export { browser } from 'wxt/browser';
            export { MigrationError, storage } from 'wxt/storage';
            export { useAppConfig } from 'wxt/utils/app-config';
            export { ContentScriptContext } from 'wxt/utils/content-script-context';
            export { createIframeUi } from 'wxt/utils/content-script-ui/iframe';
            export { createIntegratedUi } from 'wxt/utils/content-script-ui/integrated';
            export { createShadowRootUi } from 'wxt/utils/content-script-ui/shadow-root';
            export { defineAppConfig } from 'wxt/utils/define-app-config';
            export { defineBackground } from 'wxt/utils/define-background';
            export { defineContentScript } from 'wxt/utils/define-content-script';
            export { defineUnlistedScript } from 'wxt/utils/define-unlisted-script';
            export { defineWxtPlugin } from 'wxt/utils/define-wxt-plugin';
            export { injectScript } from 'wxt/utils/inject-script';
            export { InvalidMatchPattern, MatchPattern } from 'wxt/utils/match-patterns';
            export { fakeBrowser } from 'wxt/testing';
            export { startOfDay } from '../utils/time';
          }
          "
        `);
    });
  });

  describe('eslintrc', () => {
    it('"enabled: true" should output a JSON config file compatible with ESlint 8', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare({
        imports: {
          eslintrc: {
            enabled: true,
          },
        },
      });

      expect(
        await project.serializeFile('.wxt/eslintrc-auto-import.json'),
      ).toMatchSnapshot();
    });

    it('"enabled: 8" should output a JSON config file compatible with ESlint 8', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare({
        imports: {
          eslintrc: {
            enabled: 8,
          },
        },
      });

      expect(
        await project.serializeFile('.wxt/eslintrc-auto-import.json'),
      ).toMatchSnapshot();
    });

    it('"enabled: 9" should output a flat config file compatible with ESlint 9', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare({
        imports: {
          eslintrc: {
            enabled: 9,
          },
        },
      });

      expect(
        await project.serializeFile('.wxt/eslint-auto-imports.mjs'),
      ).toMatchSnapshot();
    });

    it('"enabled: false" should NOT output an ESlint config file', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare({
        imports: {
          eslintrc: {
            enabled: false,
          },
        },
      });

      expect(await project.fileExists('.wxt/eslint-auto-imports.mjs')).toBe(
        false,
      );
      expect(await project.fileExists('.wxt/eslintrc-auto-import.json')).toBe(
        false,
      );
    });

    it('should NOT output an ESlint config file by default', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare();

      expect(await project.fileExists('.wxt/eslint-auto-imports.mjs')).toBe(
        false,
      );
      expect(await project.fileExists('.wxt/eslintrc-auto-import.json')).toBe(
        false,
      );
    });

    it('should allow customizing the output', async () => {
      const project = new TestProject();
      project.addFile('entrypoints/popup.html', `<html></html>`);

      await project.prepare({
        imports: {
          eslintrc: {
            enabled: true,
            filePath: project.resolvePath('example.json'),
            globalsPropValue: 'readonly',
          },
        },
      });

      expect(await project.serializeFile('example.json')).toMatchSnapshot();
    });

    describe('Actual linting results', () => {
      async function runEslint(
        project: TestProject,
        version: boolean | 'auto' | 8 | 9,
      ) {
        project.addFile(
          'entrypoints/background.js',
          `export default defineBackground(() => {})`,
        );
        await project.prepare({
          imports: { eslintrc: { enabled: version } },
        });
        return await spawn('pnpm', ['eslint', 'entrypoints/background.js'], {
          cwd: project.root,
        });
      }

      describe('ESLint 9', () => {
        it('should have lint errors when not extending generated config', async () => {
          const project = new TestProject({
            devDependencies: {
              '@eslint/js': '9.5.0',
              eslint: '9.5.0',
            },
          });
          project.addFile(
            'eslint.config.mjs',
            `
            import eslint from "@eslint/js";

            export default [
              eslint.configs.recommended,
            ];
            `,
          );

          await expect(runEslint(project, 9)).rejects.toMatchObject({
            exitCode: 1,
            stdout: expect.stringContaining(
              "'defineBackground' is not defined",
            ),
          });
        });

        it('should not have any lint errors when configured', async () => {
          const project = new TestProject({
            devDependencies: {
              '@eslint/js': '9.5.0',
              eslint: '9.5.0',
            },
          });
          project.addFile(
            'eslint.config.mjs',
            `
            import eslint from "@eslint/js";
            import autoImports from "./.wxt/eslint-auto-imports.mjs";

            export default [
              eslint.configs.recommended,
              autoImports,
            ];
            `,
          );
          const res = await runEslint(project, 9);

          expect(res).toBeDefined();
        });
      });

      describe('ESLint 8', () => {
        it('should have lint errors when not extending generated config', async () => {
          const project = new TestProject({
            devDependencies: {
              eslint: '8.57.0',
            },
          });
          project.addFile(
            '.eslintrc',
            JSON.stringify({
              parserOptions: { ecmaVersion: 'latest', sourceType: 'module' },
              env: { es6: true },
              extends: ['eslint:recommended'],
            }),
          );

          await expect(runEslint(project, 8)).rejects.toMatchObject({
            exitCode: 1,
            stdout: expect.stringContaining(
              "'defineBackground' is not defined",
            ),
          });
        });

        it('should not have any lint errors when configured', async () => {
          const project = new TestProject({
            devDependencies: {
              eslint: '8.57.0',
            },
          });
          project.addFile(
            '.eslintrc',
            JSON.stringify({
              parserOptions: { ecmaVersion: 'latest', sourceType: 'module' },
              env: { es6: true },
              extends: [
                'eslint:recommended',
                './.wxt/eslintrc-auto-import.json',
              ],
            }),
          );
          const res = await runEslint(project, 8);

          expect(res).toBeDefined();
        });
      });
    });
  });
});
