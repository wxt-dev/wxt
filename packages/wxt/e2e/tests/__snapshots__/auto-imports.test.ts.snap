// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Auto Imports > eslintrc > "enabled: 8" should output a JSON config file compatible with ESlint 8 1`] = `
".wxt/eslintrc-auto-import.json
----------------------------------------
{
  "globals": {
    "ContentScriptContext": true,
    "InvalidMatchPattern": true,
    "MatchPattern": true,
    "MigrationError": true,
    "browser": true,
    "createIframeUi": true,
    "createIntegratedUi": true,
    "createShadowRootUi": true,
    "defineAppConfig": true,
    "defineBackground": true,
    "defineContentScript": true,
    "defineUnlistedScript": true,
    "defineWxtPlugin": true,
    "fakeBrowser": true,
    "injectScript": true,
    "storage": true,
    "useAppConfig": true
  }
}
"
`;

exports[`Auto Imports > eslintrc > "enabled: 9" should output a flat config file compatible with ESlint 9 1`] = `
".wxt/eslint-auto-imports.mjs
----------------------------------------
const globals = {
  "ContentScriptContext": true,
  "InvalidMatchPattern": true,
  "MatchPattern": true,
  "MigrationError": true,
  "browser": true,
  "createIframeUi": true,
  "createIntegratedUi": true,
  "createShadowRootUi": true,
  "defineAppConfig": true,
  "defineBackground": true,
  "defineContentScript": true,
  "defineUnlistedScript": true,
  "defineWxtPlugin": true,
  "fakeBrowser": true,
  "injectScript": true,
  "storage": true,
  "useAppConfig": true
}

export default {
  name: "wxt/auto-imports",
  languageOptions: {
    globals,
    sourceType: "module",
  },
};
"
`;

exports[`Auto Imports > eslintrc > "enabled: true" should output a JSON config file compatible with ESlint 8 1`] = `
".wxt/eslintrc-auto-import.json
----------------------------------------
{
  "globals": {
    "ContentScriptContext": true,
    "InvalidMatchPattern": true,
    "MatchPattern": true,
    "MigrationError": true,
    "browser": true,
    "createIframeUi": true,
    "createIntegratedUi": true,
    "createShadowRootUi": true,
    "defineAppConfig": true,
    "defineBackground": true,
    "defineContentScript": true,
    "defineUnlistedScript": true,
    "defineWxtPlugin": true,
    "fakeBrowser": true,
    "injectScript": true,
    "storage": true,
    "useAppConfig": true
  }
}
"
`;

exports[`Auto Imports > eslintrc > should allow customizing the output 1`] = `
"example.json
----------------------------------------
{
  "globals": {
    "ContentScriptContext": "readonly",
    "InvalidMatchPattern": "readonly",
    "MatchPattern": "readonly",
    "MigrationError": "readonly",
    "browser": "readonly",
    "createIframeUi": "readonly",
    "createIntegratedUi": "readonly",
    "createShadowRootUi": "readonly",
    "defineAppConfig": "readonly",
    "defineBackground": "readonly",
    "defineContentScript": "readonly",
    "defineUnlistedScript": "readonly",
    "defineWxtPlugin": "readonly",
    "fakeBrowser": "readonly",
    "injectScript": "readonly",
    "storage": "readonly",
    "useAppConfig": "readonly"
  }
}
"
`;
